src/v3/src/util/deviceFingerprintingUtils.ts

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2023-present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import { getUserAgent, isWindowsPhone } from './browserUtils';
+
+export const isMessageFromCorrectSource = (iframe: HTMLIFrameElement, event: MessageEvent)
+: boolean => event.source === iframe.contentWindow;
+
+// NOTE: This utility is similar to the DeviceFingerprinting.js file used for V2 authentication flows.
+export const generateDeviceFingerprint = (oktaDomainUrl: string): Promise<string> => {
+  const userAgent = getUserAgent();
+  if (!userAgent) {
+    return Promise.reject(new Error('User agent is not defined'));
+  }
+  if (isWindowsPhone(userAgent)) {
+    return Promise.reject(new Error('Device fingerprint is not supported on Windows phones'));
+  }
+
+  let timeout: NodeJS.Timeout;
+  let iframe: HTMLIFrameElement;
+  let listener: (this: Window, ev: MessageEvent) => void;
+  let msg;
+  const formElement = document.querySelector('form[data-se="o-form"]');
+  const promise = new Promise((resolve, reject) => {
+    iframe = document.createElement('iframe');
+    iframe.style.display = 'none';
+    iframe.id = 'device-fingerprint-container';
+
+    listener = (event: MessageEvent) => {
+      if (!isMessageFromCorrectSource(iframe, event)) {
+        return undefined;
```

Comment:
Should these conditions reject as well? The invocation catches and returns `undefined` anyway

Reply:
I wrote these conditions to not reject because of the case where the window can get messages from other sources in the same origin such as browser extensions (this was called out in a comment within the auth-js implementation [here](https://github.com/okta/okta-auth-js/blob/ff14212bc8e692e081621d579740eed9f50e96c5/lib/browser/fingerprint.ts#L46-L48)). This still seemed applicable here because, otherwise fingerprinting might throw an error before it can finish because of something like receiving a message from an extension

Reply:
I think you're right that the ` if (!event || !event.data || event.origin !== oktaDomainUrl)` check should reject with an error though